### PR TITLE
Link "registry" domains to learning domains

### LINKED ISSUE #1496: Support different blockchain networks (goerli, mainnet, etc)
Currently, we don't support explicitly different networks, although it can be done by combining the web3 provider configuration (e.g, `geth --goerli`) with the proper registry.

Let's try to implement a mechanism that is more robust, and that even supports to coexist more than one "nucypher network" in the same blockchain network (e.g., `testnet1` and `testnet2` in `goerli`).

### PR DESCRIPTION
**Pull request by cygnusv**
* Requires https://github.com/nucypher/ethereum-contract-registry/pull/1 (reorganization of contract registry repo)
* Reorganizes in-package registry source accordingly
* Includes an in-package registry for Testnet 2 ("_Frances_"). See https://github.com/nucypher/ethereum-contract-registry/pull/2
* Takes a stance on #1564: Defines a 1:1 mapping between "registry" and learning domains
* Fix #1496: Supports other registry domains (default was "goerli")
* Detaches Ethereum chain from registry domain (again, default was "goerli")
* `--network` is now required when created a stakeholder or ursula config
* Makes these changes effective in `nucypher stake`, `nucypher ursula` and `nucypher status`.
* Updates guides in the docs accordingly.

### CODE REVIEW

File: nucypher/cli/commands/stake.py [hunk 1]
```diff
@@ -228,12 +244,12 @@ def init_stakeholder(general_config, config_root, force, config_options):
     emitter.echo(f"Wrote new stakeholder configuration to {filepath}", color='green')
 
 
-@stake.command()
+@stake.command('list')
 @group_staker_options
 @option_config_file
 @click.option('--all', help="List all stakes, including inactive", is_flag=True)
 @group_general_config
-def list(general_config, staker_options, config_file, all):
```
Comment (KPrasch) [inline nucypher/cli/commands/stake.py]: :trumpet:

File: nucypher/config/characters.py [hunk 2]
```diff
@@ -254,6 +254,7 @@ def static_payload(self) -> dict:
         payload = dict(provider_uri=self.provider_uri,
                        poa=self.poa,
                        light=self.is_light,
+                       domains=list(self.domains),  # FIXME: 1496
```
Comment (KPrasch) [inline nucypher/config/characters.py]: Does this PR close this issue?

Comment (cygnusv) [reply nucypher/config/characters.py]: Yes, it does. I have to clean and clarify all those FIXMEs and TODOs. Perhaps open a better scoped issue, if needed.

File: nucypher/blockchain/eth/registry.py [hunk 3]
```diff
@@ -32,19 +32,44 @@
...
 
-class CanonicalRegistrySource(ABC):
+class NetworksInventory:  # TODO: Rename & relocate. See also #1564
+
+    MAINNET = 'mainnet'
+    MIRANDA = 'miranda'
+    FRANCES = 'frances'
+    CASSANDRA = 'cassandra'
+
+    UNKNOWN = 'unknown'  # TODO: Is there a better way to signal an unknown network?
```
Comment (michwill) [inline nucypher/blockchain/eth/registry.py]: constantsorrow?

File: nucypher/config/characters.py (hunk 2 above)
Comment (cygnusv) [reply nucypher/config/characters.py]: ✅

File: nucypher/blockchain/eth/registry.py [hunk 4]
```diff
@@ -32,19 +32,44 @@
 from nucypher.blockchain.eth.constants import PREALLOCATION_ESCROW_CONTRACT_NAME
 
 
-class CanonicalRegistrySource(ABC):
+class NetworksInventory:  # TODO: Rename & relocate. See also #1564
```
Comment (KPrasch) [inline nucypher/blockchain/eth/registry.py]: Consider a module `nucypher.blockchain.eth.networks`?

File: nucypher/blockchain/eth/registry.py [hunk 5]
```diff
@@ -148,23 +170,27 @@ def __init__(self, sources=None, only_primary: bool = False):
...
         Get the latest contract registry data available from a registry source chain.
         """
 
         for registry_source_class in self.sources:
             if isinstance(registry_source_class, CanonicalRegistrySource):  # i.e., it's not a class, but an instance
                 registry_source = registry_source_class
+                expected = registry_class.REGISTRY_NAME, network
+                actual = registry_source.registry_name, registry_source.network
+                if actual != expected:
+                    raise ValueError(f"(registry_name, network) should be {expected} but got {actual}")
```
Comment (KPrasch) [inline nucypher/blockchain/eth/registry.py]: Under what conditions can this occur (Do you expect it to)?

File: nucypher/utilities/sandbox/constants.py [hunk 6]
```diff
@@ -135,7 +136,7 @@ def select_test_port() -> int:
 
 MOCK_REGISTRY_FILEPATH = os.path.join(BASE_TEMP_DIR, f'{BASE_TEMP_PREFIX}mock-registry-{datetime.now().strftime(DATETIME_FORMAT)}.json')
 
-TEMPORARY_DOMAIN = ':TEMPORARY_DOMAIN:'  # for use with `--dev` node runtimes
+TEMPORARY_DOMAIN = ":TEMPORARY_DOMAIN:"  # for use with `--dev` node runtimes  # FIXME?
```
Comment (KPrasch) [inline nucypher/utilities/sandbox/constants.py]: What are you thinking to fix here?

File: docs/source/guides/staking_guide.rst [hunk 7]
```diff
@@ -140,12 +144,15 @@ Initialize a new stakeholder
 
 .. code:: bash
 
-    (nucypher)$ nucypher stake init-stakeholder --provider <PROVIDER>  --poa
+    (nucypher)$ nucypher stake init-stakeholder --provider <PROVIDER>  --poa --network <NETWORK_NAME>
```
Comment (KPrasch) [inline docs/source/guides/staking_guide.rst]: is `--poa` needed here?

File: nucypher/blockchain/eth/actors.py [hunk 8]
```diff
@@ -115,6 +115,8 @@ def __init__(self, registry: BaseContractRegistry, checksum_address: str = None)
             self.checksum_address = checksum_address  # type: str
 
         self.registry = registry
+        if domains:  # FIXME: This is horrible, but I'm forced to do it as StakeHolder config inherits from character config, which has 'domains'
```
Comment (KPrasch) [inline nucypher/blockchain/eth/actors.py]: #1580

Comment (cygnusv) [inline nucypher/blockchain/eth/actors.py]: Note to self: link to issue #1580

File: docs/source/guides/staking_guide.rst (hunk 7 above)
Comment (cygnusv) [reply docs/source/guides/staking_guide.rst]: Opened #1584

File: nucypher/utilities/sandbox/constants.py (hunk 6 above)
Comment (cygnusv) [reply nucypher/utilities/sandbox/constants.py]: Nothing, I just wasn't sure if this was an appropriate domain for tests, but I realized later that it's OK. Removed the FIXME.

File: nucypher/blockchain/eth/registry.py [hunk 9]
```diff
@@ -22,29 +22,28 @@
...
 
     name = NotImplementedError
     is_primary = NotImplementedError
 
     def __init__(self, network: str, registry_name: str, *args, **kwargs):
-        if network not in self.networks:
+        if network not in NetworksInventory.networks:
             raise ValueError(f"{self.__class__.__name__} not available for network '{network}'. "
-                             f"Only {self.networks} are allowed.")
+                             f"Valid options are: {list(NetworksInventory.networks)}")
```
Comment (jMyles) [inline nucypher/blockchain/eth/registry.py]: Using the correct `list` here, are you? ;-)

File: nucypher/blockchain/eth/contract_registry/frances/contract_registry.json [hunk 10]
```diff
@@ -0,0 +1 @@
..."},{"constant":true,"inputs":[],"name":"isOwner","outputs":[{"name":"","type":"bool"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"name":"_target","type":"address"},{"name":"_secret","type":"bytes"},{"name":"_newSecretHash","type":"bytes32"}],"name":"upgrade","outputs":[],"payable":false,"stateMutability":"nonpayable","type":"function"},{"constant":true,"inputs":[],"name":"secretHash","outputs":[{"name":"","type":"bytes32"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":true,"inputs":[],"name":"target","outputs":[{"name":"","type":"address"}],"payable":false,"stateMutability":"view","type":"function"},{"constant":false,"inputs":[{"name":"newOwner","type":"address"}],"name":"transferOwnership","outputs":[],"payable":false,"stateMutability":"nonpayable","type":"function"},{"inputs":[{"name":"_target","type":"address"},{"name":"_newSecretHash","type":"bytes32"}],"payable":false,"stateMutability":"nonpayable","type":"constructor"},{"anonymous":false,"inputs":[{"indexed":true,"name":"previousOwner","type":"address"},{"indexed":true,"name":"newOwner","type":"address"}],"name":"OwnershipTransferred","type":"event"}]]]
```
Comment (jMyles) [inline nucypher/blockchain/eth/contract_registry/frances/contract_registry.json]: https://goerli.etherscan.io/address/0xa6596Ba202f456c87aD4C745c6A35600A7fC83F9

File: nucypher/blockchain/eth/networks.py [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
+"""
+
+
+class NetworksInventory:  # TODO: See #1564
+
+    MAINNET = 'mainnet'
```
Comment (jMyles) [inline nucypher/blockchain/eth/networks.py]: Maybe some comments here explaining what each of these represent?

File: nucypher/blockchain/eth/networks.py [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    networks = tuple(__to_ethereum_chain_id.keys())
+
+    @classmethod
+    def get_ethereum_chain_id(cls, network):  # TODO: Use this (where?) to make sure we're in the right chain
+        try:
+            return cls.__to_ethereum_chain_id[network]
+        except KeyError:
+            return 1337  # TODO: what about chain id when testing?
```
Comment (jMyles) [inline nucypher/blockchain/eth/networks.py]: Is the reader supposed to know what `1337` is?  The teeth don't say: https://ethereum.stackexchange.com/a/17101

Comment (cygnusv) [reply nucypher/blockchain/eth/networks.py]: That's from our `LOCAL_CHAIN` constant in `clients.py`:
```python
LOCAL_CHAINS = {1337: "GethDev",
                5777: "Ganache/TestRPC"}
```
But it's just a placeholder. The method is unused...it's just a sketch for the future if we ever want to enforce consistency between our networks and the Ethereum network they're supposed to run upon.

**Pull request merged**